src/Browscap/Coverage/Processor.php

```diff
@@ -0,0 +1,571 @@
+                $this->coveredIds[$patternFileName] = [];
+            }
+
+            $this->coverage[$patternFileName] = $this->processFile(
+                $patternFileName,
+                file_get_contents($file->getPathname()),
+                $this->coveredIds[$patternFileName]
+            );
+        }
+    }
+
+    /**
+     * Write the coverage data in JSON format to specified filename
+     *
+     * @param string $fileName
+     */
+    public function write($fileName)
+    {
+        file_put_contents(
+            $fileName,
+            // str_replace here is to convert empty arrays into empty JS objects, which is expected by
+            // codecov.io. Owner of the service said he was going to patch it, haven't tested since
+            // Note: Can't use JSON_FORCE_OBJECT here as we **do** want arrays for the 'b' structure
+            // which FORECE_OBJECT turns into objects, breaking at least the Istanbul coverage reporter
+            str_replace('[]', '{}', json_encode($this->coverage, JSON_UNESCAPED_SLASHES))
+        );
+    }
+
+    /**
+     * Stores passed in pattern ids, grouping them by file first
+     *
+     * @param array $coveredIds
+     */
+    public function setCoveredPatternIds(array $coveredIds)
+    {
+        $this->coveredIds = $this->groupIdsByFile($coveredIds);
+    }
+
+    /**
+     * Returns the grouped pattern ids previously set
+     *
+     * @return array
+     */
+    public function getCoveredPatternIds()
+    {
+        return $this->coveredIds;
+    }
+
+    /**
+     * Process an individual file for coverage data using covered ids
+     *
+     * @param  string $file
+     * @param  string $contents
+     * @param  array  $coveredIds
+     * @return array
+     */
+    public function processFile($file, $contents, array $coveredIds)
+    {
+        $coverage = [
+            'path' => $file,
+            'statementMap' => [],
+            'fnMap' => [],
+            'branchMap' => [],
+            's' => [],
+            'b' => [],
+            'f' => [],
+        ];
+
+        $lines = explode("\n", $contents);
+
+        $u = null;
+        $d = '';
+        $c = null;
+        $p = '';
+
+        $ignores              = [];
+        $sectionBranches      = [];
+        $waitForColon         = false;
+        $collectMatchPosition = false;
+
+        $state = '';
+
+        $lexer = new Lexer();
+        $lexer->setInput($contents);
+
+        do {
+            $code    = $lexer->lex();
+            $type    = $this->symbolLookup[$code];
+            $content = $lexer->yytext;
+            $line    = $lines[$lexer->yylineno];
+
+            switch ($state) {
+                case '':
+                    if ($type === '{') {
+                        $state = 'inDivision';
+                        continue;
+                    }
+                    break;
+                case 'inDivision':
+                    if ($type === 'STRING' && $content === 'userAgents') {
```

Comment:
More constants to be defined for these (definitely `$type`, possibly `$content` depending what it's used for)